lib/grandpa/catch-up.go

```diff
@@ -0,0 +1,281 @@
+		return ErrGHOSTlessCatchUp
+	}
+
+	if err = c.verifyCatchUpResponseCompletability(prevote, msg.Hash); err != nil {
+		return fmt.Errorf("cannot verify catch-up response completability: %w", err)
+	}
+
+	// set prevotes and precommits in db
+	if err = c.grandpa.grandpaState.SetPrevotes(msg.Round, msg.SetID, msg.PreVoteJustification); err != nil {
+		return fmt.Errorf("cannot set pre votes in grandpa state: %w", err)
+	}
+
+	if err = c.grandpa.grandpaState.SetPrecommits(msg.Round, msg.SetID, msg.PreCommitJustification); err != nil {
+		return fmt.Errorf("cannot set pre commits in grandpa state: %w", err)
+	}
+
+	// update state and signal to grandpa we are ready to initiate
+	head, err := c.grandpa.blockState.GetHeader(msg.Hash)
+	if err != nil {
+		logger.Debugf("failed to process catch up response for round %d, storing the catch up response to retry", msg.Round)
+		return fmt.Errorf("cannot get header from grandpa block state: %w", err)
+	}
+
+	c.grandpa.head = head
+	c.grandpa.state.round = msg.Round
+	if c.waitingOnResponse.Load().(bool) {
+		c.catchUpResponseCh <- msg
+	}
+
+	c.waitingOnResponse.Store(false)
+
+	// resetting both response and requests
+	c.bestResponse.Store(nil)
+	c.lock.Lock()
+	c.requestsSent = make(map[peer.ID]CatchUpRequest)
+	c.lock.Unlock()
+
+	logger.Debugf("caught up to round; unpaused service and grandpa state round is %d", c.grandpa.state.round)
+	return nil
+}
+
+// verifyCatchUpResponseCompletability verifies that the pre-commit block is a descendant of, or is, the pre-voted block
+func (c *catchUp) verifyCatchUpResponseCompletability(prevote, precommit common.Hash) error {
+	if prevote == precommit {
+		return nil
+	}
+
+	// check if the current block is a descendant of prevoted block
+	isDescendant, err := c.grandpa.blockState.IsDescendantOf(prevote, precommit)
+	if err != nil {
+		return fmt.Errorf("cannot find descendance between prevote and precommit: %w", err)
+	}
+
+	if !isDescendant {
+		return ErrCatchUpResponseNotCompletable
+	}
+
+	return nil
+}
+
+func (c *catchUp) verifyPreCommitJustification(msg *CatchUpResponse) error {
+	auths := make([]AuthData, len(msg.PreCommitJustification))
+	for i, pcj := range msg.PreCommitJustification {
+		auths[i] = AuthData{AuthorityID: pcj.AuthorityID}
+	}
+
+	eqvVoters := getEquivocatoryVoters(auths)
+
+	// verify pre-commit justification
+	var count uint64
+	for idx := range msg.PreCommitJustification {
+		just := &msg.PreCommitJustification[idx]
+
+		if _, ok := eqvVoters[just.AuthorityID]; ok {
+			continue
+		}
+
+		err := verifyJustification(just, msg.Round, msg.SetID, precommit, c.grandpa.authorityKeySet())
+		if err != nil {
+			continue
+		}
+
+		if just.Vote.Hash == msg.Hash && just.Vote.Number == msg.Number {
+			count++
+		}
+	}
+
+	if count+uint64(len(eqvVoters)) < c.grandpa.state.threshold() {
+		return ErrMinVotesNotMet
+	}
+
+	return nil
+}
+
+func (c *catchUp) verifyPreVoteJustification(msg *CatchUpResponse) (common.Hash, error) {
```

Comment:
this function already exists in `lib/grandpa/message_handler.go`, why not use it?